pkg/ipmi/blob_transfer.go

```diff
@@ -0,0 +1,416 @@
+
+	// Flags
+	BMC_BLOB_OPEN_FLAG_READ  = (1 << 0)
+	BMC_BLOB_OPEN_FLAG_WRITE = (1 << 1)
+	// Blob open: bit positions 2-7 are reserved for future protocol use.
+	// Bit positions 8-15 are available for blob-specific definitions.
+	BMC_BLOB_STATE_OPEN_R       = (1 << 0)
+	BMC_BLOB_STATE_OPEN_W       = (1 << 1)
+	BMC_BLOB_STATE_COMMITTING   = (1 << 2)
+	BMC_BLOB_STATE_COMMITTED    = (1 << 3)
+	BMC_BLOB_STATE_COMMIT_ERROR = (1 << 4)
+	// Blob state: bit positions 5-7 are reserved for future protocol use.
+	// Bit positions 8-16 are available for blob-specific definitions.
+
+	// CRC options
+	REQ_CRC     = "REQ_CRC"
+	RES_CRC     = "RES_CRC"
+	NO_CRC      = "NO_CRC"
+	REQ_RES_CRC = "REQ_RES_CRC"
+
+	// OEM
+	OPENBMC = "OPENBMC"
+)
+
+// Maps OEM names to a 3 byte OEM number.
+// OENs are typically serialized as the first 3 bytes of a request body.
+var OENMap = map[string][3]uint8{
+	OPENBMC: {0xcf, 0xc2, 0x00},
+}
+
+// sendBmcCmd takes a command code, data given in little endian format, and
+// an option for cyclic redundancy checks (CRC). It constructs the request
+// and sends the command over IPMI. It receives the response, validates it,
+// and then returns the response body.
+func (i *IPMI) sendBmcCmd(code uint8, data []uint8, crcOpt CRCOption) ([]byte, error) {
+	// Initialize a buffer with the correct OEN and code.
+	oen, ok := OENMap[OPENBMC]
+	if !ok {
+		return nil, fmt.Errorf("couldn't find OEN for OPENBMC")
+	}
+
+	buf := []uint8{oen[0], oen[1], oen[2], code}
+
+	// If the request should have a CRC, derive a CRC based on the request body.
+	if crcOpt == REQ_CRC || crcOpt == REQ_RES_CRC {
+		crc := new(bytes.Buffer)
+		if err := binary.Write(crc, binary.LittleEndian, genCRC(data)); err != nil {
+			return nil, fmt.Errorf("failed to generate request CRC: ", err)
+		}
+		buf = append(buf, crc.Bytes()...)
+	}
+
+	buf = append(buf, data...)
+
+	// The request buffer should now be as follows:
+	// - 3-byte OEN
+	// - 1-byte subcommand code
+	// - (optionally) 2-byte CRC over request body in little endian format
+	// - request body in little endian format
+
+	req := &req{}
+	addr := systemInterfaceAddr{
+		addrType: _IPMI_SYSTEM_INTERFACE_ADDR_TYPE,
+		channel:  _IPMI_BMC_CHANNEL,
+		lun:      _IPMI_GGL_LUN,
+	}
+	req.addr = &addr
+	req.addrLen = uint32(unsafe.Sizeof(addr))
+	req.msg.netfn = _IPMI_GGL_NET_FN
+	req.msg.cmd = _IPMI_GGL_BLOB_CMD
+	req.msg.data = unsafe.Pointer(&buf[0])
+	req.msg.dataLen = (uint16)(len(buf))
+	req.msgid = rand.Int63()
+
+	res, err := i.sendrecv(req)
+	if err != nil {
+		return nil, err
+	}
+
+	// The response buffer is expected to be as follows:
+	// - 3-byte OEN
+	// - 2-byte CRC over response body in little endian format
+	// - response body in little endian format
+	// We verify that the OEN and CRC match the expected values.
+
+	if len(res) < OEN_LEN+1 {
+		return nil, fmt.Errorf("response too small: %d < size of OEN", len(res))
+	}
+	resOen, resBody := res[1:OEN_LEN+1], res[OEN_LEN+1:]
```

Comment:
what's in the first byte that you're chopping off here? shouldn't this be res[0:OEN_LEN] from the comment? (fix comment!)

Reply:
The first byte returned was always 0 followed the rest of the response, so I just chopped off that first 0 (similar to what the functions in ipmi.go do, they just start reading at 1 as well) - added a comment explaining that and tried to make it a little clearer!